contracts/receivables/BaseCreditPoolReceivable.sol

```diff
@@ -0,0 +1,235 @@
+        Payable
+    }
+
+    /**
+     * @dev Emitted when the owner of a receivable calls the declarePayment function
+     * @param from The address of the owner of the receivable
+     * @param to The address of the BaseCreditPool that's paid by the receivable
+     * @param tokenId The ID of the receivable token
+     * @param amount The amount that was declared paid
+     */
+    event PaymentDeclared(
+        address indexed from,
+        address indexed to,
+        uint256 indexed tokenId,
+        uint256 amount
+    );
+
+    /**
+     * @dev Constructor that sets the default admin and minter roles
+     */
+    constructor() ERC721("BaseCreditPoolReceivable", "pREC") {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MINTER_ROLE, msg.sender);
+    }
+
+    /**
+     * @dev Creates a new receivable token and assigns it to the recipient address
+     * @param recipient The address that will receive the new token
+     * @param baseCreditPool The address of the BaseCreditPool that's paid by the receivable
+     * @param paymentToken The ERC20 token used to pay the receivable
+     * @param receivableAmount The total amount of the receivable
+     * @param maturityDate The date at which the receivable becomes due
+     * @param paymentMethod The payment method used to settle the receivable
+     * @param uri The URI of the metadata associated with the receivable
+     */
+    function safeMint(
+        address recipient,
+        address baseCreditPool,
+        address paymentToken,
+        uint96 receivableAmount,
+        uint64 maturityDate,
+        PaymentMethod paymentMethod,
+        string memory uri
+    ) public onlyRole(MINTER_ROLE) {
+        (address underlyingToken, , , ) = BaseCreditPool(baseCreditPool).getCoreData();
+        require(
+            underlyingToken == paymentToken,
+            "Payment token does not match pool underlying token"
+        );
+
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(recipient, tokenId);
+
+        ReceivableInfo memory receivableInfo = ReceivableInfo(
+            BaseCreditPool(baseCreditPool),
+            paymentMethod,
+            paymentToken,
+            receivableAmount,
+            0, // Balance
+            maturityDate
+        );
+        receivableInfoMapping[tokenId] = receivableInfo;
+
+        _setTokenURI(tokenId, uri);
+    }
+
+    /**
+     * @dev Declares payment for a receivable.
+     * Only the owner of the token can declare a payment.
+     * The payment method for the receivable must be Declarative.
+     * The receivable must not already be paid in full.
+     * Emits a `PaymentDeclared` event.
+     * @param tokenId The ID of the receivable token.
+     * @param paymentAmount The amount of payment being declared.
+     */
+    function declarePayment(uint256 tokenId, uint96 paymentAmount) public {
```

Comment:
I don't know this use case very well. Is it possible that the owner didn't transfer ERC20 to pool but calls this function?